### PR TITLE
Creating role-based authorization for the query tool

### PR DESCRIPTION
**Pull request by dborgen2**
## What’s changing?

We are adding basic role and location based authorization to the Query Tool. After logging in, the logged in user must have a value for the location and also a value for their role in the NAC. If they don't we will redirect them to an unauthorized page.

In the future we will use the values of these claims to show/hide certain portions of the app based on these values, or when searching for a match we will pull in the initiating state from their location.

## Why?

Closes NAC-413

## This PR has:

- [x] **Commented source code** (_required on public classes and methods, and elsewhere as appropriate_)

- [x] **Developer documentation** (_for new build/test/API changes or complex portions of the system_)

- [x] **Automated unit tests** (_to maintain or increase level of code coverage_)

- [ ] **Changes to IAC** (_add any deployment steps that require manual intervention to the draft release notes_)

### CODE REVIEW

File: query-tool/src/Piipan.QueryTool/Pages/BasePageModel.cs [hunk 1]
```diff
@@ -12,10 +15,30 @@ public BasePageModel(IClaimsProvider claimsProvider)
...
         {
             get { return _claimsProvider.GetEmail(User); }
         }
+
+        public string State
+        {
+            get { return _claimsProvider.GetState(User); }
+        }
+
+        public string NACRole
```
Comment (bgfoshay) [inline query-tool/src/Piipan.QueryTool/Pages/BasePageModel.cs:38]: Thought: Might be worth being more generic here. Maybe just "Role"?

File: query-tool/src/Piipan.QueryTool/Pages/Error.cshtml.cs [hunk 2]
```diff
@@ -1,38 +1,22 @@
...
-using Piipan.Match.Api.Models;
+using Piipan.Shared.Authorization;
 using Piipan.Shared.Claims;
-using Piipan.Shared.Deidentification;
 
 namespace Piipan.QueryTool.Pages
 {
     public class ErrorModel : BasePageModel
     {
-        private readonly ILogger<ErrorModel> _logger;
```
Comment (bgfoshay) [inline query-tool/src/Piipan.QueryTool/Pages/Error.cshtml.cs:16]: Question: I know it wasn't used previously, but would there be any benefit to leaving the logger in and logging errors?

File: query-tool/src/Piipan.QueryTool/Pages/NotAuthorized.cshtml.cs [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+        public NotAuthorizedModel(IClaimsProvider claimsProvider)
+                          : base(claimsProvider)
+        {
+        }
+
+        [IgnoreNACAuthorization]
+        public void OnGet()
+        {
+            Message = "You do not have sufficient NAC roles or a NAC Location associated with your account";
```
Comment (bgfoshay) [inline query-tool/src/Piipan.QueryTool/Pages/NotAuthorized.cshtml.cs:18]: Issue: Might want to reword to be more generic.

File: query-tool/src/Piipan.QueryTool/appsettings.json [hunk 4]
```diff
@@ -6,9 +6,12 @@
       "Microsoft.Hosting.Lifetime": "Information"
     }
   },
-  "Claims": {
-    "Email": "email"
-  },
+    "Claims": {
+        "Email": "email",
+        "Role": "role",
+        "NACLocationPrefix": "Location-",
```
Comment (bgfoshay) [inline query-tool/src/Piipan.QueryTool/appsettings.json:12]: Issue: Maybe just LocationPrefix and RolePrefix?

File: query-tool/tests/Piipan.QueryTool.Tests/BasePageTest.cs [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+using System;
+using System.Collections.Generic;
+using System.Security.Claims;
+
+namespace Piipan.QueryTool.Tests
+{
+    public class BasePageTest
+    {
+        public static IClaimsProvider claimsProviderMock(string email = "noreply@tts.test",
+            string state = "IA", string nacRole = "Worker")
```
Comment (bgfoshay) [inline query-tool/tests/Piipan.QueryTool.Tests/BasePageTest.cs:20]: Nitpick: "Role"

File: shared/src/Piipan.Shared/Authorization/IgnoreNACAuthorizationAttribute.cs [hunk 6]
```diff
@@ -0,0 +1,9 @@
+﻿using System;
+
+namespace Piipan.Shared.Authorization
+{
+    public class IgnoreNACAuthorizationAttribute : Attribute
```
Comment (bgfoshay) [inline shared/src/Piipan.Shared/Authorization/IgnoreNACAuthorizationAttribute.cs:5]: Issue: naming again. sorry.

File: shared/src/Piipan.Shared/Claims/ClaimsOptions.cs [hunk 7]
```diff
@@ -4,5 +4,8 @@ public class ClaimsOptions
     {
         public const string SectionName = "Claims";
         public string Email { get; set; }
+        public string Role { get; set; }
+        public string NACLocationPrefix { get; set; }
```
Comment (bgfoshay) [inline shared/src/Piipan.Shared/Claims/ClaimsOptions.cs:8]: Issue: naming again. sorry.

File: shared/src/Piipan.Shared/Claims/ClaimsProvider.cs [hunk 8]
```diff
@@ -22,7 +22,32 @@ public string GetEmail(ClaimsPrincipal claimsPrincipal)
...
+            {
+                var stateClaim = identity.Claims.FirstOrDefault(c => c.Type == _options.Role && c.Value.StartsWith(_options.NACLocationPrefix));
+                if (stateClaim != null)
+                {
+                    return stateClaim.Value.Substring(_options.NACLocationPrefix.Length);
+                }
+            }
+            return null;
+        }
+        public string GetNACRole(ClaimsPrincipal claimsPrincipal)
```
Comment (bgfoshay) [inline shared/src/Piipan.Shared/Claims/ClaimsProvider.cs:40]: Issue: naming

File: shared/tests/Piipan.Shared.Tests/Claims/ClaimsProviderTests.cs [hunk 9]
```diff
@@ -52,6 +55,240 @@ public void GetEmail_NotFound()
...
+
+            // Assert
+            Assert.Equal("IA", state);
+        }
+
+        /// <summary>
+        /// Verify we can grab the NAC role from the claim
+        /// </summary>
+        [Fact]
+        public void GetNACRole()
```
Comment (bgfoshay) [inline shared/tests/Piipan.Shared.Tests/Claims/ClaimsProviderTests.cs:181]: Issue: naming

**Review by bgfoshay — COMMENTED** [review]
Mostly just naming issues. The actual content looks good and the test coverage looks great as well.

File: query-tool/src/Piipan.QueryTool/Pages/Error.cshtml.cs (hunk 2 above)
Comment (dborgen2) [reply query-tool/src/Piipan.QueryTool/Pages/Error.cshtml.cs:16]: Yes, probably a good idea. I put it back and started logging the errors.

**Pull request merged**